source/hpack/tables.d

```diff
@@ -0,0 +1,340 @@
+//module vibe.http.internal.hpack.tables;
+module HPACK.tables;
+
+import vibe.http.status;
+import vibe.http.common;
+import vibe.http.internal.http2;
+import vibe.core.log;
+
+import std.variant;
+import std.container.dlist;
+import std.traits;
+import std.meta;
+import std.range;
+import taggedalgebraic;
+
+/*
+	2.3.  Indexing Tables
+
+	HPACK uses two tables for associating header fields to indexes.  The
+	static table (see Section 2.3.1) is predefined and contains common
+	header fields (most of them with an empty value).  The dynamic table
+	(see Section 2.3.2) is dynamic and can be used by the encoder to
+	index header fields repeated in the encoded header lists.
+
+	These two tables are combined into a single address space for
+	defining index values (see Section 2.3.3).
+
+ 2.3.1.  Static Table
+
+	The static table consists of a predefined static list of header
+	fields.  Its entries are defined in Appendix A.
+
+ 2.3.2.  Dynamic Table
+
+	The dynamic table consists of a list of header fields maintained in
+	first-in, first-out order.  The first and newest entry in a dynamic
+	table is at the lowest index, and the oldest entry of a dynamic tabl
+	is at the highest index.
+	The dynamic table is initially empty.  Entries are added as each
+	header block is decompressed.
+
+	The dynamic table is initially empty.  Entries are added as each
+	header block is decompressed.
+
+	The dynamic table can contain duplicate entries (i.e., entries with
+	the same name and same value).  Therefore, duplicate entries MUST NOT
+	be treated as an error by a decoder.
+
+	The encoder decides how to update the dynamic table and as such can
+	control how much memory is used by the dynamic table.  To limit the
+	memory requirements of the decoder, the dynamic table size is
+	strictly bounded (see Section 4.2).
+
+	The decoder updates the dynamic table during the processing of a list
+	of header field representations (see Section 3.2).
+*/
+
+// wraps a header field = name:value
+struct HTTP2HeaderTableField {
+	private union HeaderValue {
+		string str;
+		string[] strarr;
+		HTTPStatus status;
+		HTTPMethod method;
+	}
+
+	string name;
+	TaggedAlgebraic!HeaderValue value;
+
+	// initializers
+	static foreach(t; __traits(allMembers, HeaderValue)) {
+		mixin("this(string n, " ~
+				typeof(__traits(getMember, HeaderValue, t)).stringof ~
+				" v) { name = n; value = v; }");
+	}
+}
+
+// fixed as per HPACK RFC
+immutable size_t STATIC_TABLE_SIZE = 61;
+
+/** static table to index most common headers
+  * fixed size, fixed order of entries (read only)
+  * cannot be updated while decoding a header block
+  */
+static immutable HTTP2HeaderTableField[size_t] StaticTable;
```

Comment:
Seems like this could be a static array instead of an AA for a little performance improvement. To make the array index checked, it could be hidden inside of an accessor function:
```
ref immutable(HTTP2HeaderTableField) getStaticTableEntry(size_t key)
{
    static immutable HTTP2HeaderTableField[STATIC_TABLE_SIZE + 1] table = [
        ...
    ];
    assert(key > 0 && key < table.length, "Invalid static table index");
    return table[key];
}
```